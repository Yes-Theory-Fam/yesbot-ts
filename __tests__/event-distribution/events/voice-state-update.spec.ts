import MockDiscord from "../../mocks/index.js";
import {
  addVoiceStateUpdateHandler,
  extractVoiceStateUpdateInfo,
  VoiceStateChange,
  VoiceStateUpdateEventHandlerOptions,
} from "../../../src/event-distribution/events/voice-state-update.js";
import { VoiceChannel } from "discord.js";
import {
  InstanceOrConstructor,
  StringIndexedHIOCTree,
} from "../../../src/event-distribution/types/hioc.js";
import {
  CommandHandler,
  DiscordEvent,
} from "../../../src/event-distribution/index.js";

describe("Voice State Update", () => {
  let mockDiscord: MockDiscord;

  beforeEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should extract information from mute events", () => {
    const secondMock = new MockDiscord();
    const oldState = mockDiscord.getVoiceState();
    const newState = { ...secondMock.getVoiceState(), mute: true };

    const result = extractVoiceStateUpdateInfo(oldState, newState);
    expect(result).toMatchSnapshot();
  });

  it("should extract information from multiple events", () => {
    const secondMock = new MockDiscord();
    const oldState = mockDiscord.getVoiceState();
    const newState = {
      ...secondMock.getVoiceState(),
      mute: true,
      channel: secondMock.getChannel() as VoiceChannel,
    };

    const result = extractVoiceStateUpdateInfo(oldState, newState);
    expect(result).toMatchSnapshot();
  });

  it("should add data to the tree", () => {
    const tree: StringIndexedHIOCTree<DiscordEvent.VOICE_STATE_UPDATE> = {};
    const options: VoiceStateUpdateEventHandlerOptions = {
      event: DiscordEvent.VOICE_STATE_UPDATE,
      changes: [VoiceStateChange.MUTED],
    };

    const ioc = {} as InstanceOrConstructor<
      CommandHandler<DiscordEvent.VOICE_STATE_UPDATE>
    >;
    addVoiceStateUpdateHandler(options, ioc, tree);
    expect(tree).toMatchSnapshot();

    const secondOptions: VoiceStateUpdateEventHandlerOptions = {
      event: DiscordEvent.VOICE_STATE_UPDATE,
      changes: [VoiceStateChange.MUTED, VoiceStateChange.JOINED],
    };

    addVoiceStateUpdateHandler(secondOptions, ioc, tree);
    expect(tree).toMatchSnapshot();
  });
});
