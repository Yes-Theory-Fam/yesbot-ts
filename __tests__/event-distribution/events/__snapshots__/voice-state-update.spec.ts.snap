// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Voice State Update should add data to the tree 1`] = `
Object {
  "MUTED": Array [
    Object {
      "ioc": Object {},
      "options": Object {
        "changes": Array [
          "MUTED",
        ],
        "event": "VOICE_STATE_UPDATE",
      },
    },
  ],
}
`;

exports[`Voice State Update should add data to the tree 2`] = `
Object {
  "JOINED": Array [
    Object {
      "ioc": Object {},
      "options": Object {
        "changes": Array [
          "MUTED",
          "JOINED",
        ],
        "event": "VOICE_STATE_UPDATE",
      },
    },
  ],
  "MUTED": Array [
    Object {
      "ioc": Object {},
      "options": Object {
        "changes": Array [
          "MUTED",
        ],
        "event": "VOICE_STATE_UPDATE",
      },
    },
    Object {
      "ioc": Object {},
      "options": Object {
        "changes": Array [
          "MUTED",
          "JOINED",
        ],
        "event": "VOICE_STATE_UPDATE",
      },
    },
  ],
}
`;

exports[`Voice State Update should extract information from multiple events 1`] = `
Array [
  Object {
    "handlerKeys": Array [
      "MUTED",
    ],
    "isDirectMessage": false,
  },
  Object {
    "handlerKeys": Array [
      "JOINED",
    ],
    "isDirectMessage": false,
  },
]
`;

exports[`Voice State Update should extract information from mute events 1`] = `
Array [
  Object {
    "handlerKeys": Array [
      "MUTED",
    ],
    "isDirectMessage": false,
  },
]
`;
